Add automatic flag-defaulting functionality.

tneymanov opened the pull request:
Create infrastructure to adding automatic flag selection and manipulation:
  - Add "Experiment" feature to flags to allow enabling/disabling certain workflows.
  - Extract manually supplied flags.
  - Add infrastructure to read, measure and manipulate VCF files to extract 5 main signals:
    - Line Count: extracted by figuring out the weight of a single record line and dividing total weight by former.
    - Sample Count: extracted by finding the last header line per file, and deriving those sample names.
    - Record Count: extracted by counting the number of lines per sample across all files.
    - Size of files: extracted from filesystem metadata information.
    - Count of files: extracted from filesystem metadata information.
  - Add functionality to set up thresholds, to control default values based on input dimensions. Added 4 thresholds of different types.

All estimations currently have relative standard error of **3%**.

Outstanding:
  - Issue: Compressed files are currently not being processed correctly, since filesystem metadata returns compressed total size, while reading lines returns the actual line weight. Need to look into it, but currently added a condition to prevent negative estimations.
  - Issue: add handling of flags with two names (eg: infer_headers/infer_undefined_headers).
  - TODO: Evaluate more optimal values and add more thresholds.
  - TODO: Modify static default values from being time efficient, to being cost/functionality focused.

PS: Added Mac OS's .DS_Store to ignore list.

kemp-google commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,217 @@
+  files.
+  """
+
+  DEFAULT_DESIRED_BUNDLE_SIZE = 64 * 1024 * 1024  # 64MB
```
Where does this value come from?

kemp-google replied:
If it's the BEAM default is it exposed at all?  i.e., could we grab it from the underlying reader instead of having it duplicated here?

kemp-google commented on gcp_variant_transforms/beam_io/vcf_estimate_io_test.py:
```diff
@@ -0,0 +1,190 @@
+
+class VcfEstimateSourceTest(unittest.TestCase):
+
+  # TODO(msaul): Replace get_full_dir() with function from utils.
```
Why is this not fixed here?

tneymanov replied:
Hmm, I missed that when copying from header tests.

However, I'm not quite certain what's being asked for here - get_full_dir() is already part of utils... I've found the PR which introduces it ( #7 ), but it doesn't have any more detail. It's not immediately clear to me which part of the current behavior is problematic either. Any idea? If not, I can just remove the comment (but I'd rather understand what was the concern two years ago...)

kemp-google replied:
Sorry means nothing to me.  Maybe Allie knows.

allieychen commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,217 @@
+      # their total size, to marginally better approximate the line count.
+      while (not header_line or
+             not header_line.strip() or header_line.startswith('##')):
+        if header_line and header_line.strip():
```
We can remove this if condition to increase readability. Also, theoretically, the white spaces should also be considered in the size.

allieychen commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,217 @@
+          header_size += len(header_line.decode('UTF-8').encode("UTF-8"))
+        header_line = file_to_read.readline()
+      if not header_line.startswith('#'):
+        raise ValueError('File {} has bad header section.'.format(file_name))
```
It can be more specific (like File has no data header line.)

allieychen commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,217 @@
+      header_size += len(header_line.decode('UTF-8').encode("UTF-8"))
+      # Use the last header line to extract samples.
+      calls = header_line.split()[8:] # Remove #CHROME..INFO mandatory fields.
+      samples = (
```
Do we need the sample list or just the size?

tneymanov commented on gcp_variant_transforms/vcf_to_bq.py:
```diff
@@ -199,6 +205,65 @@ def _shard_variants(known_args, pipeline_args, pipeline_mode):
+  temp_estimated_line_count_file_name = '-'.join(
+      [google_cloud_options.job_name,
+       datetime.now().strftime('%Y%m%d-%H%M%S'),
+       str(uuid.uuid4()),
```
Good call! Removed.

Also modified _ESTIMATE_SIZES_FILE_NAME to "estimate-sizes", instead of "estimate_sizes"

tneymanov commented:
Thanks for the review and apologies for late response. Addressed the comments.

For some reason optimize_flags.py file was restored in the commit. Deleted it and the corresponding test file.

allieychen commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -86,53 +77,69 @@ def __init__(self,
+      # TODO(#482): Find a better solution to handling compressed files.
+      all_lines_size -= header_size
+    line_size = len(first_record)
+    return float(all_lines_size) / line_size, size_in_bytes
```
It feels weird to have a count in float value.

allieychen commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -86,53 +77,69 @@ def __init__(self,
                                             splittable=False)
     self._compression_type = compression_type
 
+  def _get_header_size(self, file_to_read, file_name):
```
I am sorry, it is not quite a good idea to have these functions since each of them is kind of still relevant to the others and doing more than one thing. How about combine this and the extract_sample together, and rename it as something like _get_header_size_and_samples or just _get_header_info?

allieychen commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -20,35 +20,26 @@
   def __init__(self,
-               file_name=None,  # type: str
+               file_name,  # type: str
                estimated_line_count=None,  # type: float
```
Since they have default value (None), does it mean we may skip evaluating some of these?

allieychen commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -86,53 +77,69 @@ def __init__(self,
 
-class GetEstimates(PTransform):
+class GetEstimates(transforms.PTransform):
   """A :class:`~apache_beam.transforms.ptransform.PTransform` for reading the
```
nit: Use one line docstring or multiple lines docstring with one summary line.

allieychen commented on gcp_variant_transforms/pipeline_common.py:
```diff
@@ -131,6 +132,18 @@ def get_pipeline_mode(all_patterns, optimize_for_large_inputs=False):
 
+def get_estimates(pipeline, pipeline_mode, all_patterns):
+  # type: (beam.Pipeline, int, List[str]) -> pvalue.PCollection
+  """Creates a PCollection by reading the VCF files and deriving estimates."""
```
Since we just headers and one data line from the files, does it make sense to have a different pipeline when PipelineModes.LARGE? Or in other words, if PipelineModes.LARGE, and we use pipeline | vcf_estimate_io.GetEstimates(all_patterns[0]), is there any downside?

tneymanov replied:
well it will fail if we have input file with multiple patterns, so at least for that case we do need GetAllEstimates.

However, even if that wasn't the case, I'm not sure if there is much benefit of removing it - I mean, even if the number of header lines in the input files is small, wouldn't this still improve the speed if the number of files is very large (albeit a little)? What are we losing here, readability? We already have single vs all in other reads, so I don't think it's that confusing...

allieychen commented on gcp_variant_transforms/vcf_to_bq.py:
```diff
@@ -199,6 +203,63 @@ def _shard_variants(known_args, pipeline_args, pipeline_mode):
+                      beam.pvalue.AsSingleton(file_count),
+                      temp_estimated_line_count_file_path))
+
+  google_cloud_options.job_name = current_job_name
```
I don't think this job will change the next pipeline's job name since it will be a new instance from pipeline_options.PipelineOptions(pipeline_args)

tneymanov replied:
Hmm, good point. Removed this line and generation of current_job_name one.

allieychen commented on gcp_variant_transforms/vcf_to_bq.py:
```diff
@@ -199,6 +203,63 @@ def _shard_variants(known_args, pipeline_args, pipeline_mode):
+    raise ValueError('Exactly 5 estimates were expected in {}.'.format(
+        temp_estimated_line_count_file_path))
+
+  known_args.estimated_line_count = int(estimates[0].strip())
```
I think we can benefit from creating a new object (which has fields estimated_line_count, estimated_sample_count, etc.) for it.

tneymanov replied:
well, I can use Dimensions object that I create in the next step (when using these values to optimize the flags), but Aaron asked me to move it into the other PR for now. How about I modify that PR to create Dimensions object here instead.

Alternatively I can recreate that file but only keep Dimensions object implemented in it. Tell me if that's what you want me to do, and I'll do it.

allieychen commented on gcp_variant_transforms/transforms/extract_input_size.py:
```diff
@@ -0,0 +1,134 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A PTransform to output a PCollection of ``Variant`` records to BigQuery."""
```
update?

allieychen commented on gcp_variant_transforms/transforms/extract_input_size.py:
```diff
@@ -0,0 +1,134 @@
+            | 'GroupAllSamples' >> beam.GroupByKey())
+
+
+class GetEstimatedRecordCount(beam.PTransform):
```
I found the naming is confusion.

allieychen replied:
As discussed offline, the number of records can be simplified as number of samples * number of variants per file. But I will leave the decision to you. And also, some comments/documentation might need to help understand the `record` here.

tneymanov replied:
Added documentation to the module and to the SampleMap.

I think this is the most efficient approach, to remove any unnecessary double counting.

kemp-google replied:
I would usually refer to this as a "cell" count (rows * columns = cells)

Unfortunately, here "cells" might be overloaded.

Remind me why this value is useful?  When do we actually care how many total cells there are?  [I know we talked about it I just can't recall]

kemp-google replied:
Can we just call it Value count?  eg GetEstimatedValueCount?

tneymanov replied:
Done.

allieychen commented on gcp_variant_transforms/transforms/extract_input_size.py:
```diff
@@ -0,0 +1,134 @@
+
+@with_input_types(List[int])
+@with_output_types(int)
+class SumEstimations(beam.CombineFn):
```
Is it just summing together the elements in all PCollections? You can use beam.CombineGlobally(sum) directly. There are some other basic commonly used transforms in beam, for instance, https://github.com/apache/beam/blob/master/sdks/python/apache_beam/transforms/combiners.py

allieychen commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,219 @@
+      # TODO(#482): Find a better solution to handling compressed files.
+      all_lines_size -= header_size
+    line_size = len(first_record)
+    return float(all_lines_size) / line_size, size_in_bytes
```
It feels weird to return two irrelevant variables, especially this is something we cannot tell from the function name or docstring. I don't have a better solution, except that since `size_in_bytes = filesystems.FileSystems.match(
 [file_name])[0].metadata_list[0].size_in_bytes` is not costly, we can calculate again whenever we need it.

tneymanov replied:
Done.

tneymanov commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,219 @@
+    by subtracting header size from total size and diving it by the single line
+    size.
+    """
+    first_record = file_to_read.readline()
```
Done.

tneymanov commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,219 @@
+    return pvalue.pipeline | iobase.Read(self._source)
+
+
+def _create_vcf_estimate_source(file_pattern=None,
```
Done.

tneymanov commented on gcp_variant_transforms/beam_io/vcf_estimate_io_test.py:
```diff
@@ -0,0 +1,191 @@
+from apache_beam.testing.test_pipeline import TestPipeline
+from gcp_variant_transforms.beam_io.vcf_estimate_io import VcfEstimate
+from gcp_variant_transforms.beam_io.vcf_estimate_io import VcfEstimateSource
+from gcp_variant_transforms.beam_io.vcf_estimate_io import GetAllEstimates
```
Right, foolishly copied vcf_header_io_test :D. Done.

tneymanov commented on gcp_variant_transforms/pipeline_common.py:
```diff
@@ -32,6 +32,7 @@
 from apache_beam.runners.direct import direct_runner
 
 from gcp_variant_transforms.beam_io import vcf_header_io
+from gcp_variant_transforms.beam_io import vcf_estimate_io
```
Done.

tneymanov commented on gcp_variant_transforms/vcf_to_bq.py:
```diff
@@ -332,6 +392,13 @@ def run(argv=None):
                                                          _COMMAND_LINE_OPTIONS)
 
+  if known_args.auto_flags_experiment:
+    pipeline_mode = pipeline_common.get_pipeline_mode(
```
Good call, Done.

kemp-google commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,219 @@
+
+
+class VcfEstimateSource(filebasedsource.FileBasedSource):
+  """A source for reading VCF file estimates."""
```
"""A source that produces estimates about the number of variants and samples in a VCF file.""" ?

This makes it sound like it reads estimates

kemp-google commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,219 @@
+    # Read and skip all header lines starting with ##. Make sure to calculate
+    # their total size, to marginally better approximate the line count.
+    while (not header_line or
+           not header_line.strip() or header_line.startswith('##')):
```
Calling header_line.startsWith before strip first should be faster?

kemp-google commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,219 @@
+      header_size += len(header_line)
+      header_line = file_to_read.readline()
+    if not header_line.startswith('#'):
+      raise ValueError(('No column-defining header line  was found in file {}.'
```
extra space between `line` and `was`

kemp-google commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,219 @@
+        [file_name])[0].metadata_list[0].size_in_bytes
+    all_lines_size = size_in_bytes
+    if not isinstance(file_to_read, filesystem.CompressedFile):
+      # TODO(#482): Find a better solution to handling compressed files.
```
I wonder if we can just guess.  The ratio is probably something like 50%, +/- 5% always (complete WAGs)

kemp-google commented on gcp_variant_transforms/options/variant_transform_options.py:
```diff
@@ -597,3 +597,22 @@ def _validate_inputs(parsed_args):
+class ExperimentalOptions(VariantTransformsOptions):
+  """Options for experimental features.
+
+  All big releases should be protected behind flags. During the release, the
```
I don't think we need this commentary :) This is just what experimental flags are always used for.

kemp-google commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,223 @@
+    self.samples = samples
+    self.size_in_bytes = size_in_bytes
+
+  def __eq__(self, other):
```
Is defining equality over only one field what we want here?  Seems odd.

tneymanov replied:
Added the rest.

kemp-google commented on gcp_variant_transforms/beam_io/vcf_estimate_io.py:
```diff
@@ -0,0 +1,223 @@
+class GetAllEstimates(transforms.PTransform):
+  """Reads files until the first data line and extracts input sizes.
+
+  This transform is to be preffered over GetEstimates for large number of files.
```
preferred

tneymanov replied:
Done.

kemp-google commented on gcp_variant_transforms/options/variant_transform_options.py:
```diff
@@ -597,3 +597,15 @@ def _validate_inputs(parsed_args):
+    parser.add_argument(
+        '--auto_flags_experiment',
+        default=False,
+        help=('Experiment: Default flag to more optimal values based on '
```
How about just

'Set flag values automatically based on heuristics extracted from input files'

tneymanov replied:
Done.
